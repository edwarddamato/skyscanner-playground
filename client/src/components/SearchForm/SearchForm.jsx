import React from 'react';
import addDays from 'date-fns/add_days';
import BpkBannerAlert, { ALERT_TYPES } from 'bpk-component-banner-alert';
import BpkLabel from 'bpk-component-label';
import BpkInput, { INPUT_TYPES } from 'bpk-component-input';
import BpkSelect from 'bpk-component-select';
import BpkRadio from 'bpk-component-radio';
import BpkButton from 'bpk-component-button';
import DatePicker from './DatePicker';
import './SearchForm.scss';
import { searchPrototype } from '/utils';
import { getFlights } from '/api';
import { dispatch } from '/store';

const persons = () => {
  const options = [];
  for (let i = 0; i < 10; i++) {
    options.push(<option key={i} value={i}>{i}</option>);
  }
  return options;
};

const handleFlightSearchError = (error, componentHandler) => {
  const validationErrors = error.ValidationErrors || [];
  if (validationErrors.length > 0) {
    componentHandler(
      <div>
        <strong>Validation errors:</strong>
        {validationErrors.map((valErr, index) => {
          return (
            <div key={index}>{valErr.ParameterName ? `${valErr.ParameterName}: ` : ''}{valErr.Message}</div>
          );
        })}
      </div>
    );
  } else {
    componentHandler('Flights could not be retrieved at this moment. Please try again later.');
  }
};

/*
  TODO: Known issue - does not update app wide state with search form state.
  When menu is toggled, the search form state resets to default.
*/
class SearchForm extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      ...searchPrototype,
      originPlace: this.props.originPlace,
      destinationPlace: this.props.destinationPlace,
      outboundDate: this.props.outboundDate,
      inboundDate: this.props.inboundDate,
      cabinClass: 'Economy',
      adults: 1,
      tripType: 'return',
      exception: undefined
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleOutboundDateChange = this.handleOutboundDateChange.bind(this);
    this.handleException = this.handleException.bind(this);
    this.doSearch = this.doSearch.bind(this);
  }

  handleChange (event) {
    this.setState({ [event.target.name]: event.target.value });

    // special case for trip type - need to set inbound date to undefined if one-way
    if (event.target.name === 'tripType' &&
        event.target.value !== this.state.tripType) {
      switch (event.target.value) {
        case 'return':
          this.setState({ inboundDate: addDays(this.state.outboundDate, 1) });
          break;
        case 'oneway':
          this.setState({ inboundDate: undefined });
          break;
      }
    }
  }

  handleOutboundDateChange (event) {
    const outboundDate = event.target.value;
    this.setState({
      inboundDate: this.state.inboundDate < outboundDate ? outboundDate : this.state.inboundDate
    });
    this.handleChange(event);
  }

  handleException (exception) {
    this.setState({ exception });
  }

  doSearch () {
    dispatch.toggleLoading(true);
    dispatch.results({ searchId: '', itineraries: [] });

    const searchState = this.state;

    // this is to allow loader to load properly
    window.setTimeout(() => {
      getFlights(searchState)
        .then(response => {
          this.setState({ exception: undefined });
          dispatch.toggleMenu(false);
          dispatch.search(Object.assign({}, searchState, { state: 1 }));
          dispatch.results(response);
        })
        .catch(err => {
          handleFlightSearchError(err, this.handleException);
        })
        .then(() => {
          dispatch.toggleLoading(false);
        });
    }, 500);
  }

  render () {
    return (
      <section className="search-form">
        {
          this.state.exception
          ? <BpkBannerAlert
              message="Oops! Something went wrong when trying to fetch your flights!"
              type={ALERT_TYPES.ERROR}
              toggleButtonLabel="More Info"
            >
            {this.state.exception || ''}
            </BpkBannerAlert>
          : null
        }
        <ul className="search-form_input-list">
          <li className="search-form_input-item search-form_input-item--origin">
            <BpkLabel label="Origin" htmlFor="originPlace" />
            <BpkInput
              id="originPlace"
              type={INPUT_TYPES.TEXT}
              name="originPlace"
              value={this.state.originPlace}
              onChange={this.handleChange}
              placeholder="Airport Code (ex: LHR)"
              maxLength="4"
            />
          </li>
          <li className="search-form_input-item search-form_input-item--destination">
            <BpkLabel label="Destination" htmlFor="destinationPlace" />
            <BpkInput
              id="destinationPlace"
              type={INPUT_TYPES.TEXT}
              name="destinationPlace"
              value={this.state.destinationPlace}
              onChange={this.handleChange}
              placeholder="Airport Code (ex: MLA)"
              maxLength="4"
            />
          </li>
          <li className="search-form_input-item search-form_input-item--depart">
            <BpkLabel label="Departure Date" htmlFor="outboundDate" />
             <DatePicker
              id="outboundDate"
              name="outboundDate"
              onDateSelect={this.handleOutboundDateChange}
              date={this.state.outboundDate}
            />
          </li>
          <li className="search-form_input-item search-form_input-item--return">
            <BpkLabel label="Return Date" htmlFor="inboundDate" />
            {
              this.state.tripType === 'return'
              ? <DatePicker
                id="inboundDate"
                name="inboundDate"
                minDate={this.state.outboundDate}
                onDateSelect={this.handleChange}
                date={this.state.inboundDate}
              />
              : <BpkInput
                id="inboundDate"
                type={INPUT_TYPES.TEXT}
                name="inboundDate"
                value=""
                onChange={this.handleChange}
                placeholder="One Way"
                disabled
              />
            }
          </li>
        </ul>
        <ul className="search-form_input-list">
          <li className="search-form_input-item search-form_input-item--class">
            <BpkLabel label="Class" htmlFor="cabinClass" />
            <BpkSelect
              id="cabinClass"
              name="cabinClass"
              value={this.state.cabinClass}
              onChange={this.handleChange}>
              <option value="economy">Economy</option>
              <option value="premiumeconomy">Premium Economy</option>
              <option value="business">Business Class</option>
              <option value="first">First Class</option>
            </BpkSelect>
          </li>
          <li className="search-form_input-item search-form_input-item--type">
            <BpkLabel label="Trip Type" htmlFor="tripType" />
            <BpkRadio
              name="tripType"
              id="tripType-return"
              value="return"
              label="Return"
              onChange={this.handleChange}
              checked={this.state.tripType === 'return'}
            />
            <BpkRadio
              name="tripType"
              id="tripType-oneway"
              value="oneway"
              label="One Way"
              onChange={this.handleChange}
              checked={this.state.tripType === 'oneway'}
            />
            <BpkRadio
              name="tripType"
              id="tripType-multi"
              label="Multi-city"
              onChange={() => console.log('Not yet implemented')}
              disabled
            />
          </li>
        </ul>
        <ul className="search-form_input-list">
          <li className="search-form_input-item search-form_input-item--adults">
            <BpkLabel label="Adults" htmlFor="adults" />
            <BpkSelect
              id="adults"
              name="adults"
              value={this.state.adults.toString()}
              onChange={this.handleChange}>
              {persons()}
            </BpkSelect>
          </li>
          <li className="search-form_input-item search-form_input-item--children">
            <BpkLabel label="Children" htmlFor="children" />
            <BpkSelect
              id="children"
              name="children"
              value={this.state.children.toString()}
              onChange={this.handleChange}>
              {persons()}
            </BpkSelect>
          </li>
          <li className="search-form_input-item search-form_input-item--infants">
            <BpkLabel label="Infants" htmlFor="infants" />
            <BpkSelect
              id="infants"
              name="infants"
              value={this.state.infants.toString()}
              onChange={this.handleChange}>
              {persons()}
            </BpkSelect>
          </li>
          <li className="search-form_input-item search-form_input-item--search">
            <BpkButton large onClick={this.doSearch}>Search!</BpkButton>
          </li>
        </ul>
      </section>
    );
  }
};
SearchForm.propTypes = {
  originPlace: React.PropTypes.string.isRequired,
  destinationPlace: React.PropTypes.string.isRequired,
  outboundDate: React.PropTypes.object.isRequired,
  inboundDate: React.PropTypes.object.isRequired
};

export default SearchForm;
