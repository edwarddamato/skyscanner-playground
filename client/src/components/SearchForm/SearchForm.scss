@import "~bpk-mixins";

@mixin setViewportWidths($deskopWidth, $tabletWidth, $mobileWidth) {
  @include bpk-breakpoint-above-tablet() {
    width: $deskopWidth;
  }
  @include bpk-breakpoint-tablet-only() {
    width: $tabletWidth;
  }
  @include bpk-breakpoint-mobile() {
    width: $mobileWidth;
  }
}

.search {
  &-form {
    border: 1px solid $bpk-color-gray-50;
    @include bpk-border-radius-sm();
    margin: 2%;
    padding: 2%;

    .bpk-banner-alert {
      margin-bottom: 30px;
    }

    &_input {
      &-list {
        margin-bottom: .8em;
      }

      &-item {
        display: inline-block;

        .bpk-input {
          @include bpk-input();
        }

        @include bpk-breakpoint-tablet-only() {
          margin-bottom: 10px;
        }
        @include bpk-breakpoint-mobile() {
          margin-bottom: 5%;
        }

        &--origin {
          @include setViewportWidths($deskopWidth: 30%, $tabletWidth: 50%, $mobileWidth: 100%);

          .bpk-input {
            @include bpk-breakpoint-above-mobile() {
              @include bpk-input--docked-first-child();
            }
          }
        }
        &--destination {
          @include setViewportWidths($deskopWidth: 30%, $tabletWidth: 50%, $mobileWidth: 100%);

          .bpk-input {
            @include bpk-breakpoint-tablet-only() {
              @include bpk-input--docked-last-child();
            }
            @include bpk-breakpoint-above-tablet() {
              @include bpk-input--docked-middle-child();
            }
          }
        }
        &--depart {
          @include setViewportWidths($deskopWidth: 20%, $tabletWidth: 50%, $mobileWidth: 100%);
          
          .bpk-input {
            @include bpk-breakpoint-above-mobile() {
              @include bpk-input--docked-middle-child();
            }
          }
        }
        &--return {
          @include setViewportWidths($deskopWidth: 20%, $tabletWidth: 50%, $mobileWidth: 100%);

          .bpk-input {
            @include bpk-breakpoint-above-mobile() {
              @include bpk-input--docked-last-child();
            }
          }
        }
        &--class {
          @include setViewportWidths($deskopWidth: 50%, $tabletWidth: 30%, $mobileWidth: 100%);
        }
        &--type {
          @include setViewportWidths($deskopWidth: 50%, $tabletWidth: 70%, $mobileWidth: 100%);
          @include bpk-breakpoint-above-mobile() {
            text-align: right;
          }

          .bpk-radio {
            margin-right: 25px;
            &:last-child {
              margin-right: 0;
            }
          }
        }
        &--adults {
          @include setViewportWidths($deskopWidth: 16.5%, $tabletWidth: 16.5%, $mobileWidth: 33%);

          .bpk-select {
            @include bpk-input--docked-first-child();
          }
        }
        &--children {
          @include setViewportWidths($deskopWidth: 16.5%, $tabletWidth: 16.5%, $mobileWidth: 34%);

          .bpk-select {
            @include bpk-input--docked-middle-child();
          }
        }
        &--infants {
          @include setViewportWidths($deskopWidth: 16.5%, $tabletWidth: 16.5%, $mobileWidth: 33%);

          .bpk-select {
            @include bpk-input--docked-last-child();
          }
        }

        &--search {
          @include setViewportWidths($deskopWidth: 50%, $tabletWidth: 50%, $mobileWidth: 100%);

          text-align: right;
        }
      }
    }
  }
}