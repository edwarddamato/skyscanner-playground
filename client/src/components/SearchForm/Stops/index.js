import Stops from './Stops.jsx';
export default Stops;
