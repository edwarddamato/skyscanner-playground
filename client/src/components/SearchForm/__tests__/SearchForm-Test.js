import React from 'react';
import renderer from 'react-test-renderer';
import SearchForm from '../';

describe('SearchForm', () => {
  it('renders correctly', () => {
    const searchForm = renderer.create(
      <SearchForm
        originPlace={'MLA'}
        destinationPlace={'LGW'}
        outboundDate={new Date('2017-01-01')}
        inboundDate={new Date('2017-01-02')} />
    ).toJSON();
    expect(searchForm).toMatchSnapshot();
  });
});
