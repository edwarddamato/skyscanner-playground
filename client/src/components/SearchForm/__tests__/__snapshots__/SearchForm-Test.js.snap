// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders correctly 1`] = `
<section
  className="search-form"
>
  <ul
    className="search-form_input-list"
  >
    <li
      className="search-form_input-item search-form_input-item--origin"
    >
      <label
        className="bpk-label"
        htmlFor="originPlace"
      >
        Origin
      </label>
      <input
        className="bpk-input"
        id="originPlace"
        maxLength="4"
        name="originPlace"
        onChange={[Function]}
        placeholder="Airport Code (ex: LHR)"
        type="text"
        value="MLA"
      />
    </li>
    <li
      className="search-form_input-item search-form_input-item--destination"
    >
      <label
        className="bpk-label"
        htmlFor="destinationPlace"
      >
        Destination
      </label>
      <input
        className="bpk-input"
        id="destinationPlace"
        maxLength="4"
        name="destinationPlace"
        onChange={[Function]}
        placeholder="Airport Code (ex: MLA)"
        type="text"
        value="LGW"
      />
    </li>
    <li
      className="search-form_input-item search-form_input-item--depart"
    >
      <label
        className="bpk-label"
        htmlFor="outboundDate"
      >
        Departure Date
      </label>
      <input
        aria-atomic="true"
        aria-label="Sunday, 1st January 2017"
        aria-live="assertive"
        className="bpk-input bpk-datepicker__input"
        id="outboundDate"
        name="outboundDate_input"
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onTouchEnd={[Function]}
        type="text"
        value="Sun 1st Jan 2017"
      />
    </li>
    <li
      className="search-form_input-item search-form_input-item--return"
    >
      <label
        className="bpk-label"
        htmlFor="inboundDate"
      >
        Return Date
      </label>
      <input
        aria-atomic="true"
        aria-label="Monday, 2nd January 2017"
        aria-live="assertive"
        className="bpk-input bpk-datepicker__input"
        id="inboundDate"
        name="inboundDate_input"
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onTouchEnd={[Function]}
        type="text"
        value="Mon 2nd Jan 2017"
      />
    </li>
  </ul>
  <ul
    className="search-form_input-list"
  >
    <li
      className="search-form_input-item search-form_input-item--class"
    >
      <label
        className="bpk-label"
        htmlFor="cabinClass"
      >
        Class
      </label>
      <select
        className="bpk-select"
        id="cabinClass"
        name="cabinClass"
        onChange={[Function]}
        value="Economy"
      >
        <option
          value="economy"
        >
          Economy
        </option>
        <option
          value="premiumeconomy"
        >
          Premium Economy
        </option>
        <option
          value="business"
        >
          Business Class
        </option>
        <option
          value="first"
        >
          First Class
        </option>
      </select>
    </li>
    <li
      className="search-form_input-item search-form_input-item--type"
    >
      <label
        className="bpk-label"
        htmlFor="tripType"
      >
        Trip Type
      </label>
      <label
        className="bpk-radio"
      >
        <input
          checked={true}
          className="bpk-radio__input"
          disabled={false}
          id="tripType-return"
          name="tripType"
          onChange={[Function]}
          type="radio"
          value="return"
        />
         Return
      </label>
      <label
        className="bpk-radio"
      >
        <input
          checked={false}
          className="bpk-radio__input"
          disabled={false}
          id="tripType-oneway"
          name="tripType"
          onChange={[Function]}
          type="radio"
          value="oneway"
        />
         One Way
      </label>
      <label
        className="bpk-radio bpk-radio--disabled"
      >
        <input
          checked={false}
          className="bpk-radio__input"
          disabled={true}
          id="tripType-multi"
          name="tripType"
          onChange={[Function]}
          type="radio"
          value="tripType"
        />
         Multi-city
      </label>
    </li>
  </ul>
  <ul
    className="search-form_input-list"
  >
    <li
      className="search-form_input-item search-form_input-item--adults"
    >
      <label
        className="bpk-label"
        htmlFor="adults"
      >
        Adults
      </label>
      <select
        className="bpk-select"
        id="adults"
        name="adults"
        onChange={[Function]}
        value="1"
      >
        <option
          value={0}
        >
          0
        </option>
        <option
          value={1}
        >
          1
        </option>
        <option
          value={2}
        >
          2
        </option>
        <option
          value={3}
        >
          3
        </option>
        <option
          value={4}
        >
          4
        </option>
        <option
          value={5}
        >
          5
        </option>
        <option
          value={6}
        >
          6
        </option>
        <option
          value={7}
        >
          7
        </option>
        <option
          value={8}
        >
          8
        </option>
        <option
          value={9}
        >
          9
        </option>
      </select>
    </li>
    <li
      className="search-form_input-item search-form_input-item--children"
    >
      <label
        className="bpk-label"
        htmlFor="children"
      >
        Children
      </label>
      <select
        className="bpk-select"
        id="children"
        name="children"
        onChange={[Function]}
        value="0"
      >
        <option
          value={0}
        >
          0
        </option>
        <option
          value={1}
        >
          1
        </option>
        <option
          value={2}
        >
          2
        </option>
        <option
          value={3}
        >
          3
        </option>
        <option
          value={4}
        >
          4
        </option>
        <option
          value={5}
        >
          5
        </option>
        <option
          value={6}
        >
          6
        </option>
        <option
          value={7}
        >
          7
        </option>
        <option
          value={8}
        >
          8
        </option>
        <option
          value={9}
        >
          9
        </option>
      </select>
    </li>
    <li
      className="search-form_input-item search-form_input-item--infants"
    >
      <label
        className="bpk-label"
        htmlFor="infants"
      >
        Infants
      </label>
      <select
        className="bpk-select"
        id="infants"
        name="infants"
        onChange={[Function]}
        value="0"
      >
        <option
          value={0}
        >
          0
        </option>
        <option
          value={1}
        >
          1
        </option>
        <option
          value={2}
        >
          2
        </option>
        <option
          value={3}
        >
          3
        </option>
        <option
          value={4}
        >
          4
        </option>
        <option
          value={5}
        >
          5
        </option>
        <option
          value={6}
        >
          6
        </option>
        <option
          value={7}
        >
          7
        </option>
        <option
          value={8}
        >
          8
        </option>
        <option
          value={9}
        >
          9
        </option>
      </select>
    </li>
    <li
      className="search-form_input-item search-form_input-item--search"
    >
      <button
        className="bpk-button bpk-button--large"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Search!
      </button>
    </li>
  </ul>
</section>
`;
